Teacher forcing

Katsutoshii opened the pull request:
Removes autoregressive loop from model.call into model.call_n.
This reduces model memory requirements by a factor of roughly storm_duration / N_FLOOD_MAPS because we no longer predict the entire sequence of flood maps, which unlocks GPU training.

waltaskew commented:
Let's put together a smaller PR.

This contains more code than we want to review at a time. It also includes some code we know we don't want to merge as is, such as `metastore.py` and `featurelabelchunks.py` (those two files I rewrote as https://github.com/UrbanSystemsLab/climateiq-cnn/pull/112)

Can you put together a PR with just the changes to the model itself to start? The dataset generation we'll want to work on before reviewing & merging.

Katsutoshii commented:
> Let's put together a smaller PR.
> 
> This contains more code than we want to review at a time. It also includes some code we know we don't want to merge as is, such as `metastore.py` and `featurelabelchunks.py` (those two files I rewrote as #112)
> 
> Can you put together a PR with just the changes to the model itself to start? The dataset generation we'll want to work on before reviewing & merging.

Sure, I've reverted the extra files. The branch with e2e training colab is now at https://github.com/Katsutoshii/climateiq-cnn/tree/teacher-forcing-e2e

waltaskew commented on usl_models/usl_models/flood_ml/model.py:
```diff
@@ -103,80 +111,27 @@ def _validate_and_preprocess_data(
+        # Fit the model for this sample
+        return self._model.fit(
+            dataset, epochs=epochs, callbacks=[tb_callback] 
+            # + [es_callback] if early_stopping else []
```
Looks like this is accidentally commented out?

Katsutoshii replied:
Fixed.

tatianawu commented on usl_models/usl_models/flood_ml/model.py:
```diff
@@ -325,47 +278,54 @@ def __init__(
+        temporal: tf.Tensor = input["temporal"]
+
+        B = spatiotemporal.shape[0]
+        C = 1
```
Can we maybe add a comment saying this is the channel dimension for the spatiotemporal tensors? Unlike the other variables, there's no context to infer the variable meaning.

Katsutoshii replied:
Done.

tatianawu commented on usl_models/usl_models/flood_ml/model.py:
```diff
@@ -376,75 +336,56 @@ def forward(
         """Runs the entire autoregressive model.
 
         Args:
             inputs: A dictionary of input tensors.
```
Need to update the name here to `full_input`. Can you also explain how this differs from the `input` in the `call` method?

Katsutoshii replied:
Done.

tatianawu commented on usl_models/usl_models/flood_ml/model_params.py:
```diff
@@ -23,7 +24,7 @@ class FloodModelParams:
     # Keras models support both Optimizer objects and (valid) string names.
     # It is the user's responsibility to pass in a valid optimizer value.
-    optimizer: tf.keras.Optimizer | str = "adam"
+    optimizer: Any = dataclasses.field(default_factory=lambda: tf.keras.optimizers.Adam(learning_rate=1e-3))
```
Will this create a default Optimizer that's passed by reference to all instances if it's not overridden?

I switched to the string value in this [commit](https://github.com/UrbanSystemsLab/climateiq-cnn/pull/93/commits/62c5036ef0a2018337185f123204991e5f4bce25) to avoid possible pass-by-reference issues where the same optimizer gets shared across multiple models, which can cause downstream failures.

Katsutoshii replied:
It seems limiting to not be able to specify the parameters of Adam (learning rate etc). Using the default factor avoids these issues as long as the client does not share the Params across multiple models.

tatianawu replied:
The default value doesn't limit anything? It's just there if the user doesn't provide their own optimizer, in which they can specify whatever parameters they want. I have the string as a default to prevent any accidental pass-by-reference issues when the user _doesn't_ specify an optimizer. But your optimizer is equivalent to mine.

tatianawu commented on usl_models/usl_models/flood_ml/model.py:
```diff
@@ -376,75 +336,56 @@ def forward(
+        # Drop channels dimension.
+        return tf.squeeze(predictions, axis=-1)
+
+    def _get_temporal_window(self, temporal: tf.Tensor, t: int, n: int) -> tf.Tensor:
```
Is this a replacement for the `temporal_window_view` function in `data_utils.py`? If so, can you add more detail to the docstring explaining what this is doing? It would also be useful to add a test to make sure the behavior is WAI (and will help others understand what inputs/outputs might look like.

Katsutoshii replied:
Yes, I've added a unit test for this.

tatianawu replied:
Can you add details to the docstring explaining what the `temporal` tensor looks like when this is called?

Let's add the args and return section; I think it's better to have more detail than less here. The "n-sized window" is a bit confusing, because there are actually two temporal windows of interest in the model: the rainfall `m` and the spatiotemporal `n`. We're actually retrieving the `n` most recent (`m`-sized) rainfall windows at the current time step.

FWIW I wrote the original code and this still trips me up.

Katsutoshii replied:
Done.

tatianawu reviewed (commented):
Did you update the tests based on these changes? Pytest didn't run with the most recent update, but I'd expect some of the tests to no longer be relevant.

waltaskew commented on usl_models/usl_models/flood_ml/model.py:
```diff
@@ -103,80 +112,27 @@ def _validate_and_preprocess_data(
+    def fit(self, dataset: tf.data.Dataset, epochs: int = 1, early_stopping: int | None = None):
+        """Fit the model to the given dataset."""
+        # Create a TensorBoard callback
+        logs = "./logs/" + datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
```
what does logging into a 'logs' directory mean in the context of VertexAI training? The logs directory would be in an ephemeral docker container.
This looks like more of a debug option than a default option?

Katsutoshii replied:
Sure, I've moved this callback outside of model.fit.

Katsutoshii commented on usl_models/usl_models/flood_ml/model.py:
```diff
@@ -103,80 +112,27 @@ def _validate_and_preprocess_data(
+    def call_n(self, input: Input, n: int = 1) -> tf.Tensor:
+        return self._model.call_n(input, n=n)
+
+    def fit(self, dataset: tf.data.Dataset, epochs: int = 1, early_stopping: int | None = None):
```
Correct, I've deprecated that from _model_params to be consistent with keras API.

Katsutoshii commented on usl_models/usl_models/flood_ml/model.py:
```diff
@@ -18,6 +20,13 @@
 FloodModelParams: TypeAlias = model_params.FloodModelParams
 
 
+class Input(TypedDict):
```
I've moved this inside the class, now it's FloodModel.Input.

Katsutoshii commented on usl_models/usl_models/flood_ml/model.py:
```diff
@@ -103,80 +112,27 @@ def _validate_and_preprocess_data(
+        return self._model.call(input)
+
+    def call_n(self, input: Input, n: int = 1) -> tf.Tensor:
+        return self._model.call_n(input, n=n)
```
Updated.

Katsutoshii commented on usl_models/usl_models/flood_ml/model.py:
```diff
@@ -103,80 +112,27 @@ def _validate_and_preprocess_data(
+                                                  profile_batch='1,100')
+        es_callback = keras.callbacks.EarlyStopping(
+            monitor="val_loss", mode="min", patience=early_stopping
+        )
```
Suggestion applied.

waltaskew commented on usl_models/tests/__init__.py:
We shoudn't need all the empty `__init__.py`s in python3.
I think I'd actually prefer not to have one here in `usl_models/tests/__init__.py` because it marks it as part of the package installed by `setup.py`.

Katsutoshii replied:
I'm on latest pytest (8.2.2) and Python 3.10.14 and the tests are unable to import any of our libraries unless I include the __init__.py files.

waltaskew replied:
init files in the test modules shouldn't impact the tests' abilities to import other files
it isn't a big deal to include them, but I'm a little curious to understand the issue you're having

tatianawu replied:
Did you run `pip install` in your local environment? You'll want:

1. `pip install -r requirements.txt` to make sure you're using the same versions
2. `pip install -e .[dev]` to install the local project

waltaskew commented on usl_models/usl_models/flood_ml/dataset.py:
```diff
@@ -94,7 +172,8 @@ def generator():
         ),
     )
-    dataset = dataset.batch(batch_size)
+    if batch_size:
```
What does this branch protect against? The batch_size being 0?

Katsutoshii replied:
Yes, when batch size is zero it avoids adding an extra batch dimension, which was what caused the batch prediction error.

waltaskew replied:
Okay! A comment to that effect would be helpful.

waltaskew replied:
a test for that case would also be helpful so we don't hit this error again!

Katsutoshii replied:
Separated this into another PR.

waltaskew commented on usl_models/usl_models/flood_ml/dataset.py:
```diff
@@ -16,6 +16,88 @@
 from usl_models.flood_ml import model
 
 
+def load_dataset(
```
It looks like docstrings & comments are copy-pasted from `load_dataset_windowed`. They reference window-ing, which looks specific to `load_dataset_windowed`.

It would be helpful to explain in the docstrings when a caller should use `load_dataset` vs. `load_dataset_windowed`. Why do we need both?

We should also have a test for this function.

It may be easier to add this in another PR since it doesn't look like this function is used at present.

Katsutoshii replied:
Fixed.

tatianawu commented on usl_models/usl_models/flood_ml/model_params.py:
```diff
@@ -23,5 +24,17 @@ class FloodModelParams:
+    )
+
+
+# Used for testing.
+def test_model_params() -> FloodModelParams:
+    return FloodModelParams(
+        batch_size=4,
+        m_rainfall=3,
+        n_flood_maps=3,
+        lstm_units=32,
+        lstm_kernel_size=3,
+    )
```
Walt had initially asked for this to be removed from this file and into the `tests/` directory. Is there a reason you moved it back here?

Katsutoshii replied:
This was a merge snafu. Reverted.

tatianawu commented on usl_models/tests/flood_ml/flood_model_test.py:
```diff
@@ -115,25 +45,23 @@ def test_convlstm_call():
-    params = pytest_model_params()
+    params = model_params.test_model_params()
 
     # The FloodConvLSTM model expects the data to have been preprocessed, such
     # that it receives the full temporal inputs and a single flood map.
```
I think this comment and the docstring is slightly outdated. Can you update?

Katsutoshii replied:
Done

tatianawu commented on usl_models/tests/flood_ml/flood_model_test.py:
```diff
@@ -262,3 +146,16 @@ def test_model_checkpoint():
+    n = 5  # Window size
+    actual = flood_model.FloodConvLSTM._get_temporal_window(temporal, t=t, n=n)
+    assert actual.shape == (B, n, M)
+    expected = tf.concat([np.zeros((B, n - t, M)), temporal[:, 0:t]], axis=1)
```
This basically seems to just replicate the code from the actual method. Maybe it would be useful to create observable tensors? This has the benefit of also helping users understand what exactly the method is doing.

I'm picturing something similar to https://github.com/UrbanSystemsLab/climateiq-cnn/blob/main/usl_models/tests/atmo_ml/atmo_utils_test.py.

Katsutoshii replied:
Done.

tatianawu commented on usl_models/usl_models/flood_ml/model.py:
```diff
@@ -2,10 +2,11 @@
 import logging
-from typing import Optional, TypeAlias, TypedDict
+from typing import TypeAlias, TypedDict, List, Callable
 
 import tensorflow as tf
-from tensorflow.keras import layers
+import keras
+from keras import layers
```
Nit: Can we alphabetize?

Katsutoshii replied:
Done.

tatianawu commented on usl_models/tests/flood_ml/flood_model_test.py:
```diff
@@ -262,3 +157,24 @@ def test_model_checkpoint():
+    n = 5  # Window size
+    actual = flood_model.FloodConvLSTM._get_temporal_window(temporal, t=t, n=n)
+    assert actual.shape == (B, n, M)
+    expected = [
+        [
+            np.zeros((M)),
+            np.zeros((M)),
+            np.zeros((M)),
+            [1, 1, 1, 1, 1, 1],
+            [2, 2, 2, 2, 2, 2],
+        ],
```
Checking my understanding here. With `t = 3`, we have 2 rows of zero padding, and the data retrieved from `temporal` are rows 0-2, correct?

Could we set temporal to 1-8 instead? That row of zeros might be confusing.

tatianawu reviewed (approved):
LGTM. I ignored the dataset files. Will all of these be rebased away, or are there any I should take a look at?
